src/ray/raylet/raylet_client.h

```diff
@@ -0,0 +1,165 @@
+#ifndef RAYLET_CLIENT_H
+#define RAYLET_CLIENT_H
+
+#include <unistd.h>
+#include <mutex>
+
+#include "ray/raylet/task_spec.h"
+#include "ray/status.h"
+
+using ray::ActorID;
+using ray::JobID;
+using ray::ObjectID;
+using ray::TaskID;
+using ray::UniqueID;
+
+using MessageType = ray::protocol::MessageType;
+
+class RayletConnection {
+ public:
+  /// Connect to the local scheduler.
+  ///
+  /// \param raylet_socket The name of the socket to use to connect to the
+  ///        local scheduler.
+  /// \param worker_id A unique ID to represent the worker.
+  /// \param is_worker Whether this client is a worker. If it is a worker, an
+  ///        additional message will be sent to register as one.
+  /// \param driver_id The ID of the driver. This is non-nil if the client is a
+  ///        driver.
+  /// \return The connection information.
+  RayletConnection(const std::string &raylet_socket, int num_retries, int64_t timeout);
+
+  ~RayletConnection() { close(conn_); }
+  /// Notify the local scheduler that this client is disconnecting gracefully. This
+  /// is used by actors to exit gracefully so that the local scheduler doesn't
+  /// propagate an error message to the driver.
+  ///
+  /// \return ray::Status.
+  ray::Status Disconnect();
+  ray::Status ReadMessage(MessageType type, uint8_t *&message);
+  ray::Status WriteMessage(MessageType type,
+                           flatbuffers::FlatBufferBuilder *fbb = nullptr);
+  ray::Status AtomicRequestReply(MessageType request_type, MessageType reply_type,
+                                 uint8_t *&reply_message,
+                                 flatbuffers::FlatBufferBuilder *fbb = nullptr);
+
+ private:
+  // TODO(rkn): The io methods below should be removed.
+  int connect_ipc_sock(const std::string &socket_pathname);
+  int read_bytes(uint8_t *cursor, size_t length);
+  int write_bytes(uint8_t *cursor, size_t length);
+
+  /// File descriptor of the Unix domain socket that connects to raylet.
+  int conn_;
+  /// A mutex to protect stateful operations of the raylet client.
+  std::mutex mutex_;
+  /// A mutex to protect write operations of the local scheduler client.
+  std::mutex write_mutex_;
+};
+
+class RayletClient {
+ public:
+  /// Connect to the local scheduler.
+  ///
+  /// \param raylet_socket The name of the socket to use to connect to the raylet.
+  /// \param worker_id A unique ID to represent the worker.
+  /// \param is_worker Whether this client is a worker. If it is a worker, an
+  ///        additional message will be sent to register as one.
+  /// \param driver_id The ID of the driver. This is non-nil if the client is a
+  ///        driver.
+  /// \return The connection information.
+  RayletClient(const std::string &raylet_socket, const UniqueID &client_id,
+               bool is_worker, const JobID &driver_id, const Language &language);
+
+  ray::Status Disconnect() { return conn_->Disconnect(); };
+  /// Register with raylet.
```

Comment:
```suggestion

  /// Register with raylet.
```

applies in many places